Tests: Contract testing for the procedure AddTimeData

mmachicao opened the pull request:
Objective: Make the contract  for AddTimeData testable.

The procedure AddTimeData is stateful, so any testing beyond trivialities requires some modification in the code:

- Separation between orchestration (lock/unlock) and algorithm
- Pushing static (stateful) variables out of the procedure body

mmachicao committed:
Unit testing of procedure AddTimeData

mmachicao committed:
Unit testing of procedure AddTimeData

Pushed static variables ouf of the procedure to remove statefulness
Introduced procedure AddTimeDataAlgorithm to test the contract

mmachicao committed:
Test AddTimeDataAlgorithm

mmachicao committed:
Merge remote-tracking branch 'origin/timedata_contract_test' into timedata_contract_test

Empact commented on src/timedata.h:
```diff
@@ -75,4 +76,8 @@ int64_t GetTimeOffset();
 
+/** Functions to enable testing the contract for AddTimeData */
+int CountOffsetSamples();
+void AddTimeDataAlgorithm(const CNetAddr& ip, const int64_t nOffsetSample, std::set<CNetAddr>& knownSet, CMedianFilter<int64_t>& offsetFilter, int64_t& offset);
```
Could move these declarations to the test file, so that production code does not have access to the declarations.

Would benefit from `EXCLUSIVE_LOCKS_REQUIRED` on `AddTimeDataAlgorithm` - which if I'm not mistaken could be applied to the body of the method rather than the declaration.

mmachicao committed:
function inet_pton requires inet.h on windows

ken2812221 commented on src/test/timedata_tests.cpp:
```diff
@@ -3,6 +3,7 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 //
 #include <netaddress.h>
+#include <arpa/inet.h>
```
You could include `compat.h` instead.

mmachicao replied:
Hi. Appveyor still fails with the same message. Any other suggestion?

mmachicao replied:
Hi. A quick update. 

compat.h includes the appropriate include files for windows. Unfortunately inet_pton has a slightly differen signature in win32. Any hints how to handle this without having a windows machine at hand?. Any suggestions are welcome, since this keeps me from completing the contract tests.

Thx.

ken2812221 replied:
You can rebase your PR on #14922. That PR can fix the problem.

practicalswift commented on src/test/timedata_tests.cpp:
```diff
@@ -34,4 +36,209 @@ BOOST_AUTO_TEST_CASE(util_MedianFilter)
+}
+
+
+CNetAddr utilBuildAddress(std::string address)
```
`address` should be const ref?

practicalswift commented on src/test/timedata_tests.cpp:
```diff
@@ -34,4 +36,209 @@ BOOST_AUTO_TEST_CASE(util_MedianFilter)
+}
+
+
+void utilPreconditionIsAtLeastFiveEntriesRequired(std::string baseip, int basevalue)
```
Same here: `baseip` should be const ref?

practicalswift commented on src/test/timedata_tests.cpp:
```diff
@@ -34,4 +36,209 @@ BOOST_AUTO_TEST_CASE(util_MedianFilter)
+        std::stringstream stream;
+        stream << baseip << val;
+        std::string ip = stream.str();
+        AddTimeData(utilBuildAddress(ip.c_str()), val);
```
The call to `c_str` is redundant, right?

practicalswift commented on src/test/timedata_tests.cpp:
```diff
@@ -34,4 +36,209 @@ BOOST_AUTO_TEST_CASE(util_MedianFilter)
+}
+
+
+CNetAddr utilBuildAddress(std::string address);
```
Could be dropped if `utilBuildAddress` was defined up here instead?

mmachicao committed:
enforced code conventions and added compat.h since arpa/inet.h not available on windows

mmachicao commented on src/test/timedata_tests.cpp:
```diff
@@ -34,4 +36,209 @@ BOOST_AUTO_TEST_CASE(util_MedianFilter)
+}
+
+
+CNetAddr utilBuildAddress(std::string address);
```
Thank you for the feeback. All suggestions are done.

mmachicao committed:
compat.already included in netaddress.h shall support windows vista

mmachicao committed:
Merge branch 'master' into timedata_contract_test

mmachicao committed:
Tests for ComputeTimeSmart falsely assumed that time offset is zero

mmachicao committed:
Initialize sockaddr_in with zero before use. Use sizeof in comparison operators for CNetAddr

mmachicao committed:
util_AddTimeDataIgnoreSampleWithDuplicateIP is using ip out of range

markaw67 commented:
Win32 is labeled on my father's desktop as a virus

markaw67 commented on src/test/timedata_tests.cpp:
```diff
@@ -34,4 +35,205 @@ BOOST_AUTO_TEST_CASE(util_MedianFilter)
+    BOOST_CHECK_EQUAL(CountOffsetSamples(), samples + 1); // sample was added
+    BOOST_CHECK(GetTimeOffset() != offset);               // and new offset was computed
+}
+
```
Is this the correct address?

markaw67 commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -255,14 +256,13 @@ BOOST_FIXTURE_TEST_CASE(coin_mark_dirty_immature_credit, TestChain100Setup)
     // credit amount is calculated.
     wtx.MarkDirty();
```
Dirty huh I say he's honest

mmachicao committed:
Using memcpy to build sockaddr_in because
 compat.h does not support inet_pton with current #define _WIN32_WINNT 0x0501 and
 #define _WIN32_WINNT 0x0600 has other integration issues

mmachicao committed:
applied clang formatter

mmachicao committed:
Moved statefull tests for AddTimeData to stateless AddTimeDataAlgorithm

mmachicao committed:
improve test descriptions

mmachicao commented:
This has become illegible. 

Am replacing with: https://github.com/bitcoin/bitcoin/pull/15052

MarcoFalke commented:
Instead of closing/reopening pull request you can force push (and squash) your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits
